source/common/network/buffered_io_socket_handle_impl.cc

```diff
@@ -0,0 +1,282 @@
+                                                            uint32_t, RecvMsgOutput&) {
+  return IoSocketError::ioResultSocketInvalidAddress();
+}
+
+Api::IoCallUint64Result BufferedIoSocketHandleImpl::recvmmsg(RawSliceArrays&, uint32_t,
+                                                             RecvMsgOutput&) {
+  return IoSocketError::ioResultSocketInvalidAddress();
+}
+
+Api::IoCallUint64Result BufferedIoSocketHandleImpl::recv(void* buffer, size_t length, int flags) {
+  if (!isOpen()) {
+    return {0,
+            Api::IoErrorPtr(new IoSocketError(SOCKET_ERROR_INVAL), IoSocketError::deleteIoError)};
+  }
+  // No data and the writer closed.
+  if (pending_received_data_.length() == 0) {
+    if (read_end_stream_) {
+      return {0, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+    } else {
+      return {0, Api::IoErrorPtr(IoSocketError::getIoSocketEagainInstance(),
+                                 IoSocketError::deleteIoError)};
+    }
+  }
+  auto max_bytes_to_read = std::min(pending_received_data_.length(), length);
+  pending_received_data_.copyOut(0, max_bytes_to_read, buffer);
+  if (!(flags & MSG_PEEK)) {
+    pending_received_data_.drain(max_bytes_to_read);
+  }
+  return {max_bytes_to_read, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+}
+
+bool BufferedIoSocketHandleImpl::supportsMmsg() const { return false; }
+
+bool BufferedIoSocketHandleImpl::supportsUdpGro() const { return false; }
+
+Api::SysCallIntResult BufferedIoSocketHandleImpl::bind(Address::InstanceConstSharedPtr) {
+  return makeInvalidSyscall();
+}
+
+Api::SysCallIntResult BufferedIoSocketHandleImpl::listen(int) { return makeInvalidSyscall(); }
+
+IoHandlePtr BufferedIoSocketHandleImpl::accept(struct sockaddr*, socklen_t*) {
+  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+}
+
+Api::SysCallIntResult BufferedIoSocketHandleImpl::connect(Address::InstanceConstSharedPtr) {
+  // Buffered Io handle should always be considered as connected.
+  // Use write or read to determine if peer is closed.
+  return {0, 0};
+}
+
+Api::SysCallIntResult BufferedIoSocketHandleImpl::setOption(int, int, const void*, socklen_t) {
+  return makeInvalidSyscall();
+}
+
+Api::SysCallIntResult BufferedIoSocketHandleImpl::getOption(int, int, void*, socklen_t*) {
+  return makeInvalidSyscall();
+}
+
+Api::SysCallIntResult BufferedIoSocketHandleImpl::setBlocking(bool) { return makeInvalidSyscall(); }
+
+absl::optional<int> BufferedIoSocketHandleImpl::domain() { return absl::nullopt; }
+
+Address::InstanceConstSharedPtr BufferedIoSocketHandleImpl::localAddress() {
+  throw EnvoyException(fmt::format("getsockname failed for BufferedIoSocketHandleImpl"));
+}
+
+Address::InstanceConstSharedPtr BufferedIoSocketHandleImpl::peerAddress() {
+  throw EnvoyException(fmt::format("getsockname failed for BufferedIoSocketHandleImpl"));
+}
+
+Event::FileEventPtr BufferedIoSocketHandleImpl::createFileEvent(Event::Dispatcher& dispatcher,
```

Comment:
Merge of PR 13478 should be happening soon, you'll need to make some adjustments to switch to the new API which does help here since it transfers ownership of the file event to the IoHandle.

Reply:
adjusted